### ISSUE TITLE
Support arbitrary tunings 

### ISSUE THREAD

**Issue by infojunkie** [music-encoding/music-encoding#1167]
Right now, the [temperament data type](https://music-encoding.org/guidelines/dev/data-types/data.TEMPERAMENT.html) only accepts 4 predefined tunings. This is a serious limitation in terms of supporting world music systems, as well as Western microtonal music. 

By contrast, support for accidentals is somewhat broader and accommodates the notation of Persian, Turkish and other music systems.

What would it take to broaden the tuning / temperament support? For example, [Scala](https://www.huygens-fokker.org/scala/) supports 5900 tunings. Expecting to hard-code all these values here is counter-productive, so my suggestion instead is twofold:

1. Add a value to `data.TEMPERAMENT` that means "custom tuning"
2. Add a new attribute to `att.tuning` that allows the tuning to be specified explicitly. The simplest (but not necessarily the best) approach would be a CDATA attribute that expects Scala format.

**Comment by musicEnfanthen** [comment]
Thank you @infojunkie for bringing this up. I think that it is a great suggestion to broaden the temperament/tuning support. And also agree that hard-coding almost 6000 tunings is not a maintainable option. Do you know if there is something like a controlled vocabulary for tunings/temperaments?

I am adding this issue to the v6 milestone, since there will probably be some wish to discuss this topic and possible ways to implement such a feature in MEI.

**Comment by infojunkie** [comment]
Thanks for your reply and interest. 

I will need to dig in the literature to search for taxonomies / classifications of tunings/temperaments, so let me get back to you on that. 

However, I do want to emphasize that designing the tuning specification to be open for explicit definition of the intervals will be advantageous, because it will allow for a few interesting use cases:

a) supporting the advances made by the contemporary xenharmonic / microtonal music community
b) supporting historical tunings for musical cultures that have not been studied yet 
c) allowing a computer to faithfully play back a score by maintaining the cultural or historical accuracy of the rendition, without the need to refer to external information (i.e. the tuning intervals)

**Comment by infojunkie** [comment]
Closed by mistake :sweat_smile:

**Comment by ahankinson** [comment]
I'm wholeheartedly behind adding more temperaments to MEI, but I'm not convinced that adding them as Scala code is the way forward. It will put a requirement on compliant MEI parsers that they also support Scala code parsing as part of the spec, so it makes interoperability significantly harder. 

Since 90% of the music encoded in MEI will not specify the temperament, and 99% of the remaining 10% can probably be satisfied with an expanded formal list of temperaments, we can cover most use cases without **requiring** an external, non-MEI encoding embedded in the MEI file.

At the same time, I'm not opposed to creating an "escape hatch" for encoders to add arbitrary tuning specifications to an MEI file as part of an `<instrDef>` or something else. I think we might consider making `<instrDef>` non-MIDI specific, and allowing some child element to hold external data, which may include Scala in a CDATA block. It would be great if Scala had a formal authority URI so we might do something like:

```xml
<instrDef>
   <tuning auth.uri="https://www.huygens-fokker.org/scala/">
     <![CDATA [ ... something here ... ]]>
  </tuning>
</instrDef>
```

Or maybe something more like:

```xml
<someElementNameTBD xml:id="myTemper" auth.uri="https://www.huygens-fokker.org/scala/">
     <![CDATA [ ... Scala code here ... ]]>
</someElementNameTBD>
...
<instrDef @tune.uri="#myTemper" />
```

Is there a list of the 5,900 "tunings" that Scala supports somewhere? I couldn't find it.

**Comment by infojunkie** [comment]
The Scala tunings are available at https://www.huygens-fokker.org/scala/downloads.html#scales. Note that "scales" and "tunings" are ambiguously overlapping - maybe a music theorist could shed some light on a clear distinction.

Regarding my earlier suggestion to use a Scala tuning definition, it was only meant as an illustration of the idea, a starting point for the discussion. I don't think it's a great idea either to force consumers of MEI to parse Scala format, however simple and established it may be.

Instead, my own recommendation would be to expand the [`att.tuning`](https://music-encoding.org/guidelines/dev/attribute-classes/att.tuning.html) class spec with a new, optional attDef that explicitly states the tuning intervals as a list. In conjunction with making `@tune.temper` expandable to any text, this should be sufficiently expressive for any use case (I can think of).

I don't think expanding the `instrDef` would achieve the desired goal of describing the tuning for a whole piece, regardless of the instruments that would perform it.

**Comment by ahankinson** [comment]
> In conjunction with making @tune.temper expandable to any text, this should be sufficiently expressive for any use case (I can think of).

Could you please provide an example of what you think this might look like? I was a bit confused because you mentioned `CDATA` in your original post, but [attributes don't take CDATA](https://www.w3.org/TR/xml/#NT-AttValue) -- The `<` (and `&`) is not allowed in attributes, so you can't declare it. Also, tabs, newlines, and more than one space are not necessarily preserved as you would expect, depending on the parser.

Generally attributes are for small-ish amounts of data -- Generally fewer than 10, and typically only one or two. So if you want text of arbitrary length, or text with its own, non-XML structure, then you will probably need to wrap it as text content in an element.

> I don't think expanding the instrDef would achieve the desired goal of describing the tuning for a whole piece, regardless of the instruments that would perform it.

That's fine, we could put it on `scoreDef`.

**Comment by craigsapp** [comment]
@rettinghaus has probably thought about this a bit, so he should comment on how tuning/temperaments and should be set up (I added them to the midifile library last year for him) for use with timidity (Standard MIDI file tuning extensions may not be available in the WildWestMidi that is used with verovio).

Related to terminology, here are the standard meanings:

*  scales = a set of pitch classes that are not specific to any particular tuning.   For example C major is a scale, but it can have any tuning/temperament.

*  tuning = what frequency to assign to each note (not necessarily limited to 12-notes/octave and can be different in different octaves).     Pythagorean tuning is an example of this: the perfect fifth is the ratio 3:2, such as C-G,  you extend this (theoretically out to infinity) so that C# and D-flat have different frequencies (as well as C## and D, C### and E-flat, and so on).

* temperament =  the tuning of a fixed set of notes per octave, typically 12 notes/octave, but could be 17 notes an octave (split black keys on a keyboard), or 19 (include extra key between B-C and E-F, or 31 notes/octave.    Typically C# and D-flat are on the same key, so you choose either the C# frequency of the tuning or D-flat.

The modern description of Turkish music typically uses a theoretical 53 equal tempered chromatic scale where each step is called a Holdrian comma about the size of a 1/8 step (22.6 cents), of which the 17-pitch class scales of folk music and the 24-pitch class scales of classical music are defined in terms of (there are gaps since not all of the 53-note ET scale degrees are used).   This is an approximation of just tuning which fits well on top of 53TET:
     https://ratioscore.humdrum.org/makam

**Comment by infojunkie** [comment]
Thanks @craigsapp for clarifying - Ratioscore is a beautiful resource!
 
> Could you please provide an example of what you think this might look like?

Crudely, I am thinking of a simplification of the Makam example above, with just enough information to describe how to tune each note in the score.

```xml
<scoreDef>
   <tuning name="Turkish Classical">
      <interval>
         <ratio>1</ratio>
         <pitches><pitch pname="c" oct="4"/></pitches>
      </interval>
      <interval>
         <ratio>256/243</ratio> 
         <cents>90.57</cents>
         <pitches>
             <pitch pname="c" accid="s" oct="4"/>
             <pitch pname="d" accid="f" oct="4"/>
         </pitches>
      </interval>      
      <interval>
         <ratio>16/15</ratio>
         <cents>113.21</cents>
         <pitches>
             <pitch pname="c" accid="kms" oct="4"/>
             <pitch pname="d" accid="bf" oct="4"/>
         </pitches>
      </interval>      
      [..]
   </tuning>
</scoreDef>
```

**Comment by pe-ro** [comment]
The Standard Music Description Language (SMDL) spec, which can be found at https://xml.coverpages.org/smdl10743-1995.pdf, addresses this issue in section 8.8.  We probably don't want to copy it exactly, but there may be concepts here worth emulating.

**Comment by infojunkie** [comment]
As per in-person discussion at MEC 2026, the current suggestion is to proceed with the following:

- A `<temperamentDef>` element that defines the tuning
```xml
<temperamentDef>
      <interval ratio="1">
         <pitches><pitch pname="c"/></pitches>
      </interval>
      <interval ratio="256/243" cents="90.57">
         <pitches>
             <pitch pname="c" accid="s"/>
             <pitch pname="d" accid="f"/>
         </pitches>
      </interval>      
      <interval ratio="16/15" cents="113.21">
         <pitches>
             <pitch pname="c" accid="kms"/>
             <pitch pname="d" accid="bf"/>
         </pitches>
      </interval>      
      [..]
</temperamentDef>
```
(note: `@ratio` and `@cents` should be mutually exclusive in the real schema).

- An extension of `@tune.temper` that refers to the above temparement definition

**Comment by ahankinson** [comment]
What about this? (First a proposal, then all the problems with it below... 😁 )

```xml
<temperamentDef>
    <temperament xml:id="temper1" label="Blah blah" hz="442">
        <interval xml:id="int1" interval="700/900r" />
        <interval xml:id="int2" interval="90.57c" />
        <intervalMap>
            <pitch pname="c" interval="#int1" />
            <pitch pname="c" accid="s" interval="#int2" />
            <pitch pname="d" accid="f" interval="#int2" />
        </intervalMap>
    </temperament>
    <temperament xml:id="temper2" label="Bish Bosh">
        [..]
    </temperament>
    <temperament temper="pythagorean" xml:id="temper3">
        <intervalMap>
            <pitch ... />
        </intervalMap>
    </temperament>
    <temperament auth.temper="http://example.com/t/quarter-comma-meantone" xml:id="temper4" hz="393"/>
</temperamentDef>

<staffDef temperament="#temper1" ...>
<instrDef temperament="#temper2" ...>
```

- `temperamentDef` / `temperament`: Lets you specify more than one temperament per encoding
- `@xml:id` on temparament: address a temperament
- label: Human label on temperament
- `@interval` with units: Permit only one way of specifying an interval so we don't get conflicting values in two attributes
    - `r`: Ratio
    - `c`: Cents
    - `...`: ???
- `<intervalMap>`: Entry point for processors to map pitches to intervals without needing to traverse the children of every interval
- `<temperament>` can specify exact intervals, use a pre-defined `@temper` value, or use `@auth.temper` for an external temperament authority

Problems / questions:
- accid / accid.ges? Should we use these attributes or is there a better way to specify the pitch / accid that the interval belongs to.
- `<interval>` -- Interval is between notes. Is another name better? Maybe `<pitchDef>`?
- `cents` unit: Not an absolute unit, but a difference from a reference? 
- `@hz`: Needed? (Can be optional)
- Along with this: Remove any indication "MIDI" from `instrDef` and `instrGrp` and allow them to be a generic instrument definitions
- needed on `scoreDef`?
- part?

**Comment by infojunkie** [comment]
Thanks @ahankinson ! Comments and answers from me:

> temperamentDef / temperament: Lets you specify more than one temperament per encoding

Do we need a `temperamentDef` / `temperament` hierarchy ? I was thinking my `temperamentDef` would be playing the role of your `temperament` (i.e. multiple `temperamentDef` entries for multiple tuning definitions). What do we gain with a 2-level hierarchy?

> <intervalMap>: Entry point for processors to map pitches to intervals without needing to traverse the children of every interval

Great idea, thanks.

> @interval with units: Permit only one way of specifying an interval so we don't get conflicting values in two attributes
 
Yes, much better. In tuning literature, only ratios and cents are used.

> accid / accid.ges? Should we use these attributes or is there a better way to specify the pitch / accid that the interval belongs to.

Using `accid` specifially is important imo because we are mapping from visual accidentals (as notated in the score) to a tuning degree. I see a precedent in [`tuning/course`](https://music-encoding.org/guidelines/v5/content/tablature.html#tablature-tuning).

> <interval> -- Interval is between notes. Is another name better? Maybe <pitchDef>?

Pitches and intervals are duals, so both would work to denote tuning degrees in my opinion. The problem I see with `pitch(Def)` is that the term "pitch" is already used for named pitches C,D,E,... So it would become an ambiguous term, which is not ideal.

> @hz: Needed? (Can be optional)

Not needed imo, `@tune.Hz` is already there as reference frequency.

> needed on scoreDef?
> part?

This tuning definition extends the existing [`att.tuning` class](https://music-encoding.org/guidelines/v5/attribute-classes/att.tuning.html) which is available on `scoreDef`, `layerDef`, `staffDef`, and I suggest keeping the structure as is. 

> Along with this: Remove any indication "MIDI" from instrDef and instrGrp and allow them to be a generic instrument definitions

Looking at [`instrDef`](https://music-encoding.org/guidelines/v5/elements/instrDef.html), I see it's tightly bound to MIDI so I would suggest to keep it out of this present spec to avoid opening a can of worms.

That's why I am suggesting to extend the definition of `@tune.temper` to accept an `xml:id` that refers to `temperamentDef` / `temperament`.

**Comment by ahankinson** [comment]
> Do we need a temperamentDef / temperament hierarchy ? I was thinking my temperamentDef would be playing the role of your temperament (i.e. multiple temperamentDef entries for multiple tuning definitions).

`<temperamentDef>` serves as the parent element for the list. This keeps all the temperament definitions in one place, and makes the modelling easy (allow only `temperamentDef` in a specific location, e.g., `<meiHead>` or whatever.)

This is also useful if you want to define all your temperaments in a single MEI file, and then reference that externally in other files. So if you put your list of temperaments in `foobar.mei`, then you can use `foobar.mei#temper1` in any URI field. (And indeed, `http://example.com/foobar.mei#temper1`)

> Using accid specifially is important imo because we are mapping from visual accidentals (as notated in the score) to a tuning degree. I see a precedent in [tuning/course](https://music-encoding.org/guidelines/v5/content/tablature.html#tablature-tuning).

In MEI terms, I think we're creating a **gestural** mapping to a performed pitch, no? (See the "Musical Domains" explanation here: https://music-encoding.org/guidelines/v5/content/introduction.html#musicalDomains).

I was thinking out loud here, though -- I think pname / accid is likely the best approach, but was just opening it up to see if there were any other suggestions.

> Not needed imo, `@tune.Hz` is already there as reference frequency.

The reason I suggested it is that it is the bit of information needed to actually calculate the performed frequency of a given temperament. If you know a specific instrument (say an organ) is always tuned to 446, and you are encoding a score that was almost certainly performed on that organ (say, a Bach piece) then you can actually calculate the actual frequencies for each note based on the known reference frequency. It wouldn't be required, but could be useful? 

> is that the term "pitch" is already used for named pitches C,D,E

That's kinda the reason for the question -- The way I see this is that we are actually specifying the performed frequency of a "C", rather than assuming 12TET as we do now. 

I don't know enough about this, though, but it seems a bit off that we, e.g., map a `<pitch>` (D flat) to a single `<interval>` definition. In other words, why aren't we mapping **two** pitches to a single interval? There's just something a bit off on the modelling here.

**Comment by infojunkie** [comment]
> <temperamentDef> serves as the parent element for the list.

Sounds good, I'm convinced.

> In MEI terms, I think we're creating a gestural mapping to a performed pitch, no?

I think we're mapping from the _written_ accidentals in the score to the performed pitch. In my microtonal PR, It would have probably been more ontologically accurate to first convert the written accidentals to gestural accidentals, but because there's no 1:1 mapping between written and gestural accidentals, we'd be risking losing information if e.g. ACCIDENTAL_WRITTEN_1qf (which is mapped to ACCIDENTAL_GESTURAL_fu) was encountered in the score.

> The reason I suggested it is that it is the bit of information needed to actually calculate the performed frequency

Now that I'm thinking about it, we actually need 3 pieces of information about the reference pitch, corresponding to Ableton ASCL's `@ABL REFERENCE_PITCH 4 9 440.0`: octave, MIDI key index within the octave, frequency. So you're right, it would be good to encode these 3 attributes (with appropriate defaults):

`<temperament @hz="440" @oct="4" @midi="9">`

> we are actually specifying the performed frequency of a "C"

Yes, we are. I am convinced.

--- 

There's also one case I didn't mention: there are tuning definitions that necessitate multiple accidentals per pitch, e.g. accidentals in the [Extended Helmholtz-Ellis set](https://w3c-cg.github.io/smufl/latest/tables/extended-helmholtz-ellis-accidentals-just-intonation.html) are expected to be combined for precise tuning. Just like [MEI (and Verovio) support multiple accidentals per note](https://www.verovio.org/test-suite.xhtml?cat=accid) (`accid-013.mei`), the same should be supported in this tuning definition. It could look like this:

```xml
        <intervalMap>
            <pitch pname="c" interval="#int2">
                <accid glyph.name="accidentalCombiningRaise53LimitComma" glyph.auth="smufl"/>
                <accid glyph.name="accidentalRaiseOneUndecimalQuartertone" glyph.auth="smufl"/>
            </pitch>
            <pitch pname="d" accid="f" interval="#int2" />
        </intervalMap>
```

**Comment by ahankinson** [comment]
> I think we're mapping from the written accidentals in the score to the performed pitch

MEI is a bit more specific in what we mean by **written**. There is the **logical** definition, which is the musical structure, and then there is the **visual** definition, which is the actual graphical representation. This is separated because a `<note pname="d" accid="s" />` may not actually have a written (that is, visual) accidental if, for example, D# is in the key signature, or if a D# has occurred previously in the same measure. The pitch is still a D#, but there may not actually be an accidental visible. 

It's not always cut-and-dry, and can get a bit convoluted at times, but we try to make sure we know in which musical domain(s) we're operating with these types of changes. So this change would likely map to affecting the logical / gestural domain, and not really affecting the visual domain at all. 

For:

` <accid glyph.name="accidentalCombiningRaise53LimitComma" glyph.auth="smufl"/>`

I wouldn't do this... these are visual definitions (i.e., "use this glyph when rendering it"). Better to expand the values available through `data.ACCIDENTAL.WRITTEN` https://music-encoding.org/guidelines/dev/mei-all/data-types/data.ACCIDENTAL.WRITTEN.html

("Written" here doesn't just mean visual, confusingly...)

> .. octave, MIDI key index within the octave .. 

could you expand on why these are needed? I'm not sure we have the concept of the "MIDI key index within the octave" -- it would be better to specify the exact MIDI note number if we need it, not the number relative to the octave.

**Comment by infojunkie** [comment]
> So this change would likely map to affecting the logical / gestural domain, and not really affecting the visual domain at all.

I'd say we're mapping from the logical domain to tuning degrees.

> I wouldn't do this... these are visual definitions (i.e., "use this glyph when rendering it").

What do you suggest to support SMuFL accidentals that are not present in the `data.ACCIDENTAL.WRITTEN` set ? There are ~480 accidental symbols in SMuFL.

> could you expand on why these are needed? [..] it would be better to specify the exact MIDI note number if we need it, not the number relative to the octave.

Agreed. A MIDI key  (note number) should be sufficient.

**Comment by infojunkie** [comment]
> I wouldn't do this... these are visual definitions (i.e., "use this glyph when rendering it").

Maybe we can rename the attribute itself to something like `accid.smufl` instead of `glyph.name` to clearly show we don't care about the visual domain, but about the semantic meaning of this SMuFL symbol?

**Comment by ahankinson** [comment]
But SMuFL isn’t at all concerned with semantics and only visuals, so that wouldn’t work.

**Comment by infojunkie** [comment]
> But SMuFL isn’t at all concerned with semantics and only visuals, so that wouldn’t work.

In principle, I agree. In the case of microtonal accidentals, however, we need to be able to express the full set of accidentals that music systems of the world use. The only source of such information is SMuFL at the moment. Would MEI be interested to reproduce and maintain the hundreds of such accidentals?

**Comment by ahankinson** [comment]
> Would MEI be interested to reproduce and maintain the hundreds of such accidentals?

That's part of what we're trying to figure out here. 

There's no point in having a value like `accidentalCombiningRaise53LimitComma` in MEI if it doesn't "mean" something other than what to draw on a page. If all you want to do is draw something, then the `glyph.*` attributes already exist. Clearly we want to do something more here, though.

In principle there is no reason why MEI can't take in all the accidentals as part of its datatypes. That's not necessarily a problem, but there are two issues I see:

1. There are a number of accidentals that are duplicates, or near duplicates, e.g., E270, E2C6. There would likely need to be a review of all the accidentals to see which ones are visual duplicates and which ones are semantic duplicates (that is, the same meaning but with a different glyph). I suspect that this would reduce the overall number. Also we wouldn't need things like parentheses or combining characters. 
2. Naming and defining the effect. SMuFL gives an indication of where it comes from, and sometimes uses the effect as part of the name, but (for example) I'm not sure what the sounding effect of `accidentalDoubleSharpOneArrowDown` (E2C4) is. Each accidental, if it's to have an actual tuning effect on playback, would need to have its effect on the note defined. 

Turning to your example above, there are a few questions I have with your current modelling that I've been trying to work through.  I'm working from this:

```
<intervalMap>
    <pitch pname="c" interval="#int2">
           <accid glyph.name="accidentalCombiningRaise53LimitComma" glyph.auth="smufl"/>
           <accid glyph.name="accidentalRaiseOneUndecimalQuartertone" glyph.auth="smufl"/>
    </pitch>
    <pitch pname="d" accid="f" interval="#int2" />
</intervalMap>
```

1. I wouldn't use `<accid>` here. As I see it, this is a definition of the accidental class, rather than an "instance" of an accidental. Using the same tag for both would likely require building two different models for the same tag name, something that occurs rarely in MEI, and we try to avoid it. 
2. The `<intervalMap>`, as I see it, maps a `pname (+ accid)` value to a specific frequency realization of this. So with the `<pitch pname="c" interval="#int2">` and child accidentals, the issue here is that I'm not sure what the note instance used here would look like. In other words, if I write `<note pname="c">` in my encoding, would that draw a note with two accidentals based on this example?
3 .In music notation, accidentals are cumulative: If I write a flat, it lowers it a half-step. If I write two flats, it lowers it a whole step. Are the accidental definitions above also cumulative? If I combine two accidentals (as in the example) do they each have a similar cumulative effect? Where / how is each effect defined? 

In short, I'm not entirely convinced that the `<temperament>` needs to concern itself with multiple **drawn** accidentals. The way I see it is that a definition like this `<pitch pname="d" accid="f" interval="#intN">` is used to identify how `<note pname="d" accid="f" />` is _performed_. 

Also, the pitch would have to have the same effect on:

```
<note pname="d">
   <accid accid="f" />
</note>
```

That's not to say that I don't see the value in having something like you suggest; I'm just not sure how that would work in combination with all the other bits of encoding. That's why I'm suggesting we keep it in the performance domain for now, since bringing in visuals confuses the issue (at least, it does for me.)

**Comment by infojunkie** [comment]
Maybe we can anchor this discussion with a concrete example:

<img width="2100" height="928" alt="Image" src="https://github.com/user-attachments/assets/375cf4ef-415e-4b53-9302-e048a0cb2389" />

This score uses sagittal accidentals which use small ratios for their note tunings. We are not interested in the visual rendering of the score per se, but we want to be able to say: 

Give a score with those accidentals, produce a MIDI file that tunes the notes to the given ratios.

How are the score accidentals encoded? Using SMuFL accidentals. In MusicXML, the first E looks like:

```xml
<note>
<pitch>
  <step>E</step>
  <octave>4</octave>
</pitch>
<accidental smufl="accSagittal5CommaDown">other</accidental>
[..]
```
and in MEI, it looks like:

```xml
<note xml:id="yvoc4qn" oct="4" pname="e">
  <accid xml:id="zbrpqcl" accid="n" accid.ges="n" glyph.auth="smufl" glyph.name="accSagittal5CommaDown" />
</note>
```

Now we need to tell MEI that this EaccSagittal5CommaDown is tuned to 5/4 the root. This is where the tuning comes in:

```xml
<temperament>
  <pitchDef interval="5/4r" xml:id="#pd1" />
  <pitchMap>
    <pitch name="e" SOME.ATTR="accSagittal5CommaDown">
  </pitchMap>
</temperament>
```

Given this info, the MEI processor knows to produce a MIDI tuning with a given frequency when it encounters the note EaccSagittal5CommaDown.

---


Given this example, we can answer your questions:

> Each accidental, if it's to have an actual tuning effect on playback, would need to have its effect on the note defined.

The effect is given in the tuning definition `<temperament>`. 

What happens today is that there's an implicit tuning where C = 0 cents, Cs = Df = 100 cents, D = 200 cents, etc.

> I wouldn't use <accid> here. 

Yes, best to avoid naming collisions.

>  In other words, if I write <note pname="c"> in my encoding, would that draw a note with two accidentals based on this example?

It's the other way around. The score rendering logic remains unchanged. It's just the auditory rendering that gets affected, so `<note pname="c">` without any accidental would try to match this unaccidented note in the tuning. In the example above, the E note is encoded in the score with its (smufl) accidental. 

> In music notation, accidentals are cumulative: If I write a flat, it lowers it a half-step. If I write two flats, it lowers it a whole step. Are the accidental definitions above also cumulative? If I combine two accidentals (as in the example) do they each have a similar cumulative effect? Where / how is each effect defined?

Yes, the combination of accidentals is always cumulative. The difference of accidentals that are explicitly marked as "combining" is that they are never expected to be found alone with a note: they are always expected to combine with another accidental - thus why we expect to find entries with multiple accidentals in the tuning. 

---

Reading your questions and my answers, I am wondering if you are expecting the accidentals to have a default effect (e.g. s = +100c, f = -100c, accSagittal5CommaDown = +5.4r, etc.) as opposed to the approach described here which explicitly lists each combination of note+accidental that is expected to be found in the score. The issue is that we cannot always assume that we're starting from C = 100 cents, D = 200 cents, etc. For full microtonal expressivity, we need to be able to assign an arbitrary tuning degree to any note+accidental combination.

**Comment by ahankinson** [comment]
OK, let's work through this.

```xml
<note xml:id="yvoc4qn" oct="4" pname="e">
  <accid xml:id="zbrpqcl" accid="n" accid.ges="n" glyph.auth="smufl" glyph.name="accSagittal5CommaDown" />
</note>
```

This is saying "Play an E4 natural, but instead of showing the natural accidental, show the SMuFL glyph for `accSagittal5CommaDown`. There is no effect on the pitch; That is, I would expect it to be played back at 329.63 Hz (E4). 

Now if we add temperament modifications:

```xml
<temperament>
  <pitchDef interval="5/4r" xml:id="#pd1" />
  <pitchMap>
    <pitch name="e" SOME.ATTR="accSagittal5CommaDown">
  </pitchMap>
</temperament>
```

The `accSagittal5CommaDown` has no effect, since it's not an accidental, it's a glyph. If we want to affect the above note example, I would expect:

```xml
<pitch name="e" accid="n" interval="#pd1">
```

Which "redefines" E natural to have a ratio of 5/4. 

However, if we specifically want to define the effect of `accSagittal5CommaDown` then we might need something like this, assuming `accSagittal5CommaDown` comes into MEI as an accidental code of `s5cd`. 

```xml
<note xml:id="yvoc4qn" oct="4" pname="e">
  <accid xml:id="zbrpqcl" accid="s5cd" accid.ges="s5cd" />
</note>
```

Or, since this is a single note, this also works:

```xml
<note xml:id="yvoc4qn" oct="4" pname="e" accid="s5cd">
```

Then in temperament:

```xml
<pitch name="e" accid="s5cd" interval="#pd1">
```

(`s5cd` is just a placeholder, we can figure out a naming scheme later... )

The tricky bit comes when you have multiple accidentals. Looking at the B in the fourth measure. As best I can tell it uses these two accidentals:
 - `accSagittal5CommaDown` --> `s5cd`
 - `accSagittalFlat7CDown` -> `sf7cd`

Thus the MEI: 

```xml
<note pname="b" oct="4">
    <accid accid="s5cd" />
    <accid accid="sf7cd" />
</note>
```

But then we run into a problem with `<pitch>`. We could allow `@accid` on `<pitch>` to contain a list of values:

```xml
<pitch pname="b" accid="s5cd sf7cd" interval="#pd2" />
```

The upside is that this is nice and compact, but the downside is that we would need to define the interval for every single note, and every single possible combination of N accidentals, since we don't define the **effect** of an `s5cd` or an `sf7cd`. 

And that's where I'm a bit stuck. I could imagine:

```xml
<pitch pname="b" interval="#pd2" />
<pitch accid="s5cd" interval="#pd3" />
<pitch accid="sf7cd" interval="#pd4" />
```

And then make sure the `@interval` has a way to describe the accidental effect, i.e., defining the cents a saggital 5th comma down "removes" from a base pitch?

But then I'm not sure how that would work with temperaments where a D# and an Eb are actually different notes, since the effect of a "flat" or a "sharp" is not universal across all notes -- The flat on an E might have a different distance than a flat on a B. 

At this point, I think I should revisit the Scala definitions and see if there is anything there that we can borrow.

**Comment by infojunkie** [comment]
Thanks, I love the fact that you're willing to consider extending the set of accidentals. 

(As a aside, I would wish for a way to specify in MEI that all SMuFL accidentals are automatically part of the spec, instead of creating a parallel listing and playing catch-up and/or introducing inconsistencies).

Assuming the needed accidentals are in the spec, then I suggest the `pitch/accid` spec is designed as a mirror of the `note/accid` spec, but using a different name than `accid`:

```xml
<pitch pname="b" paccid="s5cd" interval="#int1"/>
```

and 

```xml
<pitch pname="b" interval="#int2">
  <paccid paccid="s5cd" />
  <paccid paccid="sf7cd" />
</pitch>
```

**Comment by lpugin** [comment]
I am not completely convinced by the idea of extending the accidental types to match the values we have in SMuFL. As far as I can see, with the current support for custom tuning, we have about 450 different accidentals. I can imagine that some of these might be grouped under a common accidental type, but at least in theory, each of them should be distinguishable. So having a way to refer externally to SMuFL in tunings, in the same way we do it for `accid`, might still be necessary.

**Comment by ahankinson** [comment]
> As a aside, I would wish for a way to specify in MEI that all SMuFL accidentals are automatically part of the spec, instead of creating a parallel listing and playing catch-up and/or introducing inconsistencies

A case of being "careful of what you wish for." 

SMuFL only defines a visual glyph, it doesn't define how to interpret it. MEI operates at the semantic level first, which means we need to define how to interpret it first, and then how to draw it comes later. 

Looking at this:

```xml
<pitch pname="b" interval="#int2">
  <paccid paccid="s5cd" />
  <paccid paccid="sf7cd" />
</pitch>
```

I guess where I'm stuck is: What are those accidentals (`<paccid>`) doing? We've already defined the exact interval (by pointing to the `#int2` definition), and presumably `#int2` takes into account the pitch-modifyoing effect of both of the accidentals. 

To hopefully illustrate a bit further, when we then want to *use* the definition, does it look like this? 

```xml
<note pname"b">
    <accid accid="s5cd" />
    <accid accid="sf7cd" />
</note>
```

If so, what's the point of having the `<paccid>` definitions as child elements? If it was to simply match pitch instances to notes, then a shorter definition would be possible, something like:

```xml
<pitch pname="b" paccid="s5cd sf7cd" interval="#int2">
...
```

But then, like I said, where do we stop? Do we need:

```xml
<pitch pname="b" interval="#intA">
<pitch pname="b" paccid="s5cd" interval="#intB">
<pitch pname="b" paccid="sf7cd" interval="#intC">
<pitch pname="b" paccid="s5cd sf7cd" interval="#intD">
<pitch pname="c" interval="#intE">
<pitch pname="c" paccid="s5cd" interval="#intF">
<pitch pname="c" paccid="sf7cd" interval="#intG">
<pitch pname="c" paccid="s5cd sf7cd" interval="#intH">
<pitch pname="d" interval="#intJ">
<pitch pname="d" paccid="s5cd" interval="#intK">
<pitch pname="d" paccid="sf7cd" interval="#intL">
<pitch pname="d" paccid="s5cd sf7cd" interval="#intM">
... etc ... 
```

+++ all the other accidentals and pnames in order to arrive at a "fully defined" tuning map?

**Comment by infojunkie** [comment]
> [..] but at least in theory, each of them should be distinguishable. So having a way to refer externally to SMuFL in tunings, in the same way we do it for accid, might still be necessary.

That's also my opinion here.

Maybe we can consider injecting SMuFL accidentals into the score using something like the following:

```xml
<accidDef glyph.name="accSagittal5CommaDown" glyph.auth="smufl" accid="s5cd" />
<accidDef glyph.name="accSagittalFlat7CDown" glyph.auth="smufl" accid="sf7cd" />
```

which would then be available to use in the score's `accid` and `accid.ges`, and also referenced in the `temperament/pitch` declarations above.

> [..] If it was to simply match pitch instances to notes, then a shorter definition would be possible, something like:

If it's fine to supply a space-separated list that the application needs to parse instead of an XML list, then so be it.

> Do we need: [list of all pitch mappings] +++ all the other accidentals and pnames in order to arrive at a "fully defined" tuning map?

Absolutely, that's precisely what SCL+KBM and ASCL do.

=====

One more detail regarding the expected effect of accidentals on notes, and why not all accidentals can be considered to be additive to the regular 12-tet pitches.

Accidentals such as the Johnston set, the Extended Helmholtz-Ellis set and various Sagittal sets are all based on [Pythagorean (Just Intonation) tuning](https://en.wikipedia.org/wiki/Pythagorean_tuning) rather that equal temperament. So even though the value of each accidental is constant regardless of pitch, they apply to just intonation pitches rather 12-tet pitches. 

To see this in action, let's do a calculation:

- Looking at the first table in the Wikipedia article on Pythagorean tuning, the unison is D and E sits at 3.91 cents diff from 12-tet

<img width="894" height="330" alt="Image" src="https://github.com/user-attachments/assets/c0de98bd-18c9-4ba6-a433-56b3611b06e9" />

- At the [HEJI calculator](https://hejicalc.plainsound.org/), set the Reference Pitch to D natural
- We will apply the accidental [accidentalFlatOneArrowDown](https://w3c-cg.github.io/smufl/latest/tables/extended-helmholtz-ellis-accidentals-just-intonation.html) to D and E respectively:
- Select DaccidentalFlatOneArrowDown and see the MuseScore tuning to be -135.19 cents

<img width="1492" height="675" alt="Image" src="https://github.com/user-attachments/assets/f7d9b28f-9e9c-4bf0-8cc5-10c958d2af43" />

- Select EaccidentalFlatOneArrowDown and see the MuseScore tuning to be -131.28 cents

<img width="1490" height="660" alt="Image" src="https://github.com/user-attachments/assets/72168a62-939e-49a7-8d97-4fcf1a8693ea" />

- Verify that `(-131.28) - (-135.19) = 3.91`, which corresponds to the difference between the reference and E natural in JI space.
